File: packages/storage/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/storage/FirebaseStorageHelper.java
package io.capawesome.capacitorjs.plugins.firebase.storage;

import static com.google.firebase.storage.StorageException.ERROR_BUCKET_NOT_FOUND;
import static com.google.firebase.storage.StorageException.ERROR_CANCELED;
import static com.google.firebase.storage.StorageException.ERROR_INVALID_CHECKSUM;
import static com.google.firebase.storage.StorageException.ERROR_NOT_AUTHENTICATED;
import static com.google.firebase.storage.StorageException.ERROR_NOT_AUTHORIZED;
import static com.google.firebase.storage.StorageException.ERROR_OBJECT_NOT_FOUND;
import static com.google.firebase.storage.StorageException.ERROR_PROJECT_NOT_FOUND;
import static com.google.firebase.storage.StorageException.ERROR_QUOTA_EXCEEDED;
import static com.google.firebase.storage.StorageException.ERROR_RETRY_LIMIT_EXCEEDED;

import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.google.firebase.storage.StorageException;
import com.google.firebase.storage.StorageMetadata;
import java.util.Iterator;
import org.json.JSONException;
import org.json.JSONObject;

public class FirebaseStorageHelper {

    @Nullable
    public static StorageMetadata buildStorageMetadata(@Nullable JSObject metadata) throws JSONException {
        if (metadata == null) {
            return null;
        }
        String cacheControl = metadata.getString("cacheControl");
        String contentDisposition = metadata.getString("contentDisposition");
        String contentEncoding = metadata.getString("contentEncoding");
        String contentLanguage = metadata.getString("contentLanguage");
        String contentType = metadata.getString("contentType");
        JSONObject customMetadata = metadata.getJSObject("customMetadata");

        StorageMetadata.Builder builder = new StorageMetadata.Builder();
        if (cacheControl != null) {
            builder = builder.setCacheControl(cacheControl);
        }
        if (contentDisposition != null) {
            builder = builder.setContentDisposition(contentDisposition);
        }
        if (contentEncoding != null) {
            builder = builder.setContentEncoding(contentEncoding);
        }
        if (contentLanguage != null) {
            builder = builder.setContentLanguage(contentLanguage);
        }
        if (contentType != null) {
            builder = builder.setContentType(contentType);
        }
        if (customMetadata != null) {
            Iterator<String> keys = customMetadata.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                Object value = customMetadata.get(key);
                if (value instanceof String) {
                    builder = builder.setCustomMetadata(key, (String) value);
                }
            }
        }

        return builder.build();
    }

    @Nullable
    public static String createErrorCode(@Nullable Exception exception) {
        if (exception == null) {
            return null;
        } else if (exception instanceof StorageException) {
            return convertErrorCodeToString(((StorageException) exception).getErrorCode());
        }
        return null;
    }

    private static String convertErrorCodeToString(int code) {
        return switch (code) {
            case ERROR_BUCKET_NOT_FOUND -> "storage/bucket-not-found";
            case ERROR_CANCELED -> "storage/canceled";
            case ERROR_INVALID_CHECKSUM -> "storage/invalid-checksum";
            case ERROR_NOT_AUTHENTICATED -> "storage/unauthenticated";
            case ERROR_NOT_AUTHORIZED -> "storage/unauthorized";
            case ERROR_OBJECT_NOT_FOUND -> "storage/object-not-found";
            case ERROR_PROJECT_NOT_FOUND -> "storage/project-not-found";
            case ERROR_QUOTA_EXCEEDED -> "storage/quota-exceeded";
            case ERROR_RETRY_LIMIT_EXCEEDED -> "storage/retry-limit-exceeded";
            default -> "storage/unknown";
        };
    }

    private static String snakeToKebabCase(String snakeCase) {
        return snakeCase.replaceAll("_+", "-").toLowerCase();
    }
}

File: packages/storage/ios/Plugin/FirebaseStorageHelper.swift
import Foundation
import FirebaseStorage
import Capacitor

public class FirebaseStorageHelper {
    public static func buildStorageMetadata(_ metadata: JSObject) -> StorageMetadata {
        let storageMetadata = StorageMetadata()
        if let cacheControl = metadata["cacheControl"] as? String {
            storageMetadata.cacheControl = cacheControl
        }
        if let contentDisposition = metadata["contentDisposition"] as? String {
            storageMetadata.contentDisposition = contentDisposition
        }
        if let contentEncoding = metadata["contentEncoding"] as? String {
            storageMetadata.contentEncoding = contentEncoding
        }
        if let contentLanguage = metadata["contentLanguage"] as? String {
            storageMetadata.contentLanguage = contentLanguage
        }
        if let contentType = metadata["contentType"] as? String {
            storageMetadata.contentType = contentType
        }
        if let customMetadata = metadata["customMetadata"] as? JSObject {
            storageMetadata.customMetadata = self.createHashMapFromJSObject(customMetadata)
        }
        return storageMetadata
    }

    private static func createHashMapFromJSObject(_ object: JSObject) -> [String: String] {
        var map: [String: String] = [:]
        for key in object.keys {
            if let value = object[key] as? String {
                map[key] = value
            }
        }
        return map
    }

    public static func createErrorCode(error: Error?) -> String? {
        if let error = error as NSError? {
            return convertErrorCodeToString(errorCode: error.code)
        }
        return nil
    }

    private static func convertErrorCodeToString(errorCode: Int) -> String? {
        let errorCodes: [Int: String] = [
            StorageErrorCode.bucketMismatch.rawValue: "storage/bucket-mismatch",
            StorageErrorCode.bucketNotFound.rawValue: "storage/bucket-not-found",
            StorageErrorCode.cancelled.rawValue: "storage/cancelled",
            StorageErrorCode.downloadSizeExceeded.rawValue: "storage/download-size-exceeded",
            StorageErrorCode.internalError.rawValue: "storage/internal-error",
            StorageErrorCode.invalidArgument.rawValue: "storage/invalid-argument",
            StorageErrorCode.nonMatchingChecksum.rawValue: "storage/non-matching-checksum",
            StorageErrorCode.objectNotFound.rawValue: "storage/object-not-found",
            StorageErrorCode.pathError.rawValue: "storage/path-error",
            StorageErrorCode.projectNotFound.rawValue: "storage/project-not-found",
            StorageErrorCode.quotaExceeded.rawValue: "storage/quota-exceeded",
            StorageErrorCode.retryLimitExceeded.rawValue: "storage/retry-limit-exceeded",
            StorageErrorCode.unauthenticated.rawValue: "storage/unauthenticated",
            StorageErrorCode.unauthorized.rawValue: "storage/unauthorized",
            StorageErrorCode.unknown.rawValue: "storage/unknown"
        ]

        return errorCodes[errorCode]
    }
}
